File: app/backend/src/lambda/course/create.rs
use aws_lambda_events::event::apigw::{ApiGatewayProxyRequest, ApiGatewayProxyResponse};
use aws_lambda_events::encodings::Body;
use lambda_runtime::{Error, LambdaEvent};
use serde_json::json;
use tracing::{error, info};
use http::HeaderMap;
use surrealdb::sql::Thing;

use crate::common::auth;
use crate::common::db;
use crate::common::error::AppError;
use crate::models::course::{Course, CourseCreateRequest};
use crate::models::user::UserRole;

/// Lambda handler for course creation
pub async fn handler(event: LambdaEvent<ApiGatewayProxyRequest>) -> Result<ApiGatewayProxyResponse, Error> {
    let request = event.payload;

    // Parse request body
    let course_request = match request.body {
        Some(body) => match serde_json::from_str::<CourseCreateRequest>(&body) {
            Ok(req) => req,
            Err(err) => {
                error!("Failed to parse course creation request: {}", err);
                return Ok(AppError::Validation(format!("Invalid request format: {}", err)).into());
            }
        },
        None => {
            return Ok(AppError::Validation("Missing request body".to_string()).into());
        }
    };

    // Validate request fields
    if course_request.title.is_empty() || course_request.description.is_empty() {
        return Ok(AppError::Validation("Title and description are required".to_string()).into());
    }

    // Extract and validate authorization token
    let token = match request.headers.get("Authorization") {
        Some(auth_header) => {
            let auth_str = auth_header.to_str().unwrap_or_default();
            let parts: Vec<&str> = auth_str.split_whitespace().collect();
            if parts.len() != 2 || parts[0].to_lowercase() != "bearer" {
                return Ok(AppError::Authentication("Invalid authorization format".to_string()).into());
            }
            parts[1]
        }
        None => {
            return Ok(AppError::Authentication("Missing authorization header".to_string()).into());
        }
    };

    // Validate token and extract claims
    let claims = match auth::validate_token(token) {
        Ok(claims) => claims,
        Err(err) => {
            return Ok(err.into());
        }
    };

    // Check if user has educator or admin role
    if claims.role != "educator" && claims.role != "admin" {
        return Ok(AppError::Authorization("Only educators and admins can create courses".to_string()).into());
    }

    // Connect to database
    let db = match db::get_db_client().await {
        Ok(client) => client,
        Err(err) => {
            error!("Failed to connect to database: {}", err);
            return Ok(
                AppError::Internal("Failed to connect to the database".to_string()).into(),
            );
        }
    };
    
    // Create user reference
    let educator = Thing::from(("user", &*claims.sub));

    // Create course object
    let new_course = Course {
        id: None,
        title: course_request.title,
        description: course_request.description,
        difficulty: course_request.difficulty,
        tags: course_request.tags,
        educator,
        created_at: Some(chrono::Utc::now()),
        updated_at: Some(chrono::Utc::now()),
        is_published: course_request.is_published.unwrap_or(false),
        thumbnail: course_request.thumbnail_url,
        duration_hours: 0.0, // Default duration, will be updated as content is added
    };

    // Insert course into database
    let created_course: Result<Option<Course>, _> = db
        .query("CREATE course CONTENT $data RETURN *")
        .bind(("data", new_course))
        .await
        .and_then(|mut res| res.take::<Option<Course>>(0));

    let course = match created_course {
        Ok(Some(course)) => course,
        Ok(None) => {
            return Ok(AppError::Internal("Course created but not returned".to_string()).into());
        }
        Err(err) => {
            error!("Failed to create course: {}", err);
            return Ok(AppError::Database(err).into());
        }
    };

    // Create successful response
    let response_body = json!({
        "message": "Course created successfully",
        "course": course
    });

    let mut headers = HeaderMap::new();
    headers.insert("Content-Type", "application/json".parse().unwrap());

    Ok(ApiGatewayProxyResponse {
        status_code: 201, // Created
        headers,
        multi_value_headers: HeaderMap::new(),
        body: Some(Body::from(response_body.to_string())),
        is_base64_encoded: false,
    })
} 

File: app/backend/nodejs-lambda/src/handlers/courses/create.js
/**
 * Course creation handler
 */
const { success, error } = require('../../utils/response');
const { createCourse } = require('../../utils/db');
const { v4: uuidv4 } = require('uuid');
const { requireRole } = require('../../middleware/auth');

/**
 * Handle course creation request
 * @param {object} event - API Gateway Lambda proxy event
 * @returns {object} Lambda proxy response
 */
const handler = async (event) => {
  try {
    // Verify user is an educator
    try {
      event = await requireRole('educator')(event);
    } catch (err) {
      return error(403, 'Authorization Error', err.message);
    }

    // Parse request body
    const body = JSON.parse(event.body || '{}');
    
    // Validate required fields
    const requiredFields = ['title', 'description', 'difficulty', 'category'];
    const missingFields = requiredFields.filter(field => !body[field]);
    
    if (missingFields.length > 0) {
      return error(400, 'Bad Request', `Missing required fields: ${missingFields.join(', ')}`);
    }
    
    // Generate course ID
    const courseId = uuidv4();
    
    // Create course object
    const course = {
      id: courseId,
      title: body.title,
      description: body.description,
      difficulty: body.difficulty,
      category: body.category,
      educator: event.user.id, // Set educator from authenticated user
      tags: body.tags || [],
      thumbnail: body.thumbnail || null,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
      is_published: body.is_published ? 'true' : 'false', // Convert boolean to string for DynamoDB
      duration_hours: body.duration_hours || 0,
      students: [],
      rating: 0,
      reviews: [],
      price: body.price || 0
    };
    
    // Create course in DynamoDB
    await createCourse(course);
    
    // Return success response
    return success(201, { 
      message: 'Course created successfully',
      course
    });
  } catch (err) {
    console.error('Error in course creation handler:', err);
    return error(500, 'Internal Server Error', err.message || 'An unexpected error occurred');
  }
};

module.exports = { handler }; 